File: src/hyrr/materials.py
"""Bridge between py-mat materials and HYRR isotopic resolution."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from hyrr.db import DatabaseProtocol
    from hyrr.models import Element


# Periodic table data for formula parsing (Z and standard atomic weights)
# Only need symbol -> Z mapping for formula parsing
SYMBOL_TO_Z: dict[str, int] = {
    "H": 1, "He": 2, "Li": 3, "Be": 4, "B": 5, "C": 6, "N": 7, "O": 8,
    "F": 9, "Ne": 10, "Na": 11, "Mg": 12, "Al": 13, "Si": 14, "P": 15,
    "S": 16, "Cl": 17, "Ar": 18, "K": 19, "Ca": 20, "Sc": 21, "Ti": 22,
    "V": 23, "Cr": 24, "Mn": 25, "Fe": 26, "Co": 27, "Ni": 28, "Cu": 29,
    "Zn": 30, "Ga": 31, "Ge": 32, "As": 33, "Se": 34, "Br": 35, "Kr": 36,
    "Rb": 37, "Sr": 38, "Y": 39, "Zr": 40, "Nb": 41, "Mo": 42, "Tc": 43,
    "Ru": 44, "Rh": 45, "Pd": 46, "Ag": 47, "Cd": 48, "In": 49, "Sn": 50,
    "Sb": 51, "Te": 52, "I": 53, "Xe": 54, "Cs": 55, "Ba": 56, "La": 57,
    "Ce": 58, "Pr": 59, "Nd": 60, "Pm": 61, "Sm": 62, "Eu": 63, "Gd": 64,
    "Tb": 65, "Dy": 66, "Ho": 67, "Er": 68, "Tm": 69, "Yb": 70, "Lu": 71,
    "Hf": 72, "Ta": 73, "W": 74, "Re": 75, "Os": 76, "Ir": 77, "Pt": 78,
    "Au": 79, "Hg": 80, "Tl": 81, "Pb": 82, "Bi": 83, "Po": 84, "At": 85,
    "Rn": 86, "Fr": 87, "Ra": 88, "Ac": 89, "Th": 90, "Pa": 91, "U": 92,
}

# Standard atomic weights for mass<->atom fraction conversion
STANDARD_ATOMIC_WEIGHT: dict[str, float] = {
    "H": 1.008, "He": 4.003, "Li": 6.941, "Be": 9.012, "B": 10.81,
    "C": 12.01, "N": 14.01, "O": 16.00, "F": 19.00, "Ne": 20.18,
    "Na": 22.99, "Mg": 24.31, "Al": 26.98, "Si": 28.09, "P": 30.97,
    "S": 32.07, "Cl": 35.45, "Ar": 39.95, "K": 39.10, "Ca": 40.08,
    "Sc": 44.96, "Ti": 47.87, "V": 50.94, "Cr": 52.00, "Mn": 54.94,
    "Fe": 55.85, "Co": 58.93, "Ni": 58.69, "Cu": 63.55, "Zn": 65.38,
    "Ga": 69.72, "Ge": 72.63, "As": 74.92, "Se": 78.97, "Br": 79.90,
    "Kr": 83.80, "Rb": 85.47, "Sr": 87.62, "Y": 88.91, "Zr": 91.22,
    "Nb": 92.91, "Mo": 95.95, "Ru": 101.1, "Rh": 102.9, "Pd": 106.4,
    "Ag": 107.9, "Cd": 112.4, "In": 114.8, "Sn": 118.7, "Sb": 121.8,
    "Te": 127.6, "I": 126.9, "Xe": 131.3, "Cs": 132.9, "Ba": 137.3,
    "La": 138.9, "Ce": 140.1, "Pr": 140.9, "Nd": 144.2, "Sm": 150.4,
    "Eu": 152.0, "Gd": 157.3, "Tb": 158.9, "Dy": 162.5, "Ho": 164.9,
    "Er": 167.3, "Tm": 168.9, "Yb": 173.0, "Lu": 175.0, "Hf": 178.5,
    "Ta": 180.9, "W": 183.8, "Re": 186.2, "Os": 190.2, "Ir": 192.2,
    "Pt": 195.1, "Au": 197.0, "Hg": 200.6, "Tl": 204.4, "Pb": 207.2,
    "Bi": 209.0, "Th": 232.0, "Pa": 231.0, "U": 238.0,
}


def parse_formula(formula: str) -> dict[str, int]:
    """Parse a chemical formula into element counts.

    Examples:
        "MoO3" -> {"Mo": 1, "O": 3}
        "H2O" -> {"H": 2, "O": 1}
        "Al2O3" -> {"Al": 2, "O": 3}
        "Cu" -> {"Cu": 1}

    Args:
        formula: Chemical formula string

    Returns:
        Dict mapping element symbol to atom count
    """
    pattern = r"([A-Z][a-z]?)(\d*)"
    elements: dict[str, int] = {}
    for match in re.finditer(pattern, formula):
        symbol = match.group(1)
        count = int(match.group(2)) if match.group(2) else 1
        if symbol:  # skip empty matches
            elements[symbol] = elements.get(symbol, 0) + count
    return elements


def formula_to_mass_fractions(formula: str) -> dict[str, float]:
    """Convert chemical formula to elemental mass fractions.

    Args:
        formula: Chemical formula (e.g., "MoO3")

    Returns:
        Dict mapping element symbol to mass fraction (sums to 1.0)
    """
    elem_counts = parse_formula(formula)
    total_mass = 0.0
    masses: dict[str, float] = {}
    for symbol, count in elem_counts.items():
        mass = count * STANDARD_ATOMIC_WEIGHT[symbol]
        masses[symbol] = mass
        total_mass += mass
    return {sym: m / total_mass for sym, m in masses.items()}


def mass_to_atom_fractions(mass_fractions: dict[str, float]) -> dict[str, float]:
    """Convert mass fractions to atom fractions.

    x_i = (w_i / M_i) / sum(w_j / M_j)

    Args:
        mass_fractions: Dict of element symbol -> mass fraction

    Returns:
        Dict of element symbol -> atom fraction (sums to 1.0)
    """
    moles: dict[str, float] = {}
    total = 0.0
    for symbol, w in mass_fractions.items():
        m = w / STANDARD_ATOMIC_WEIGHT[symbol]
        moles[symbol] = m
        total += m
    return {sym: m / total for sym, m in moles.items()}


def resolve_element(
    db: DatabaseProtocol,
    symbol: str,
    enrichment: dict[int, float] | None = None,
) -> Element:
    """Resolve an element with natural or enriched isotopic composition.

    Args:
        db: Database for natural abundance lookup
        symbol: Element symbol (e.g., "Mo")
        enrichment: Optional dict of A -> fractional abundance overrides.
            If provided, must sum to ~1.0. Replaces natural abundances entirely.

    Returns:
        Element with isotopic composition
    """
    from hyrr.models import Element

    Z = SYMBOL_TO_Z[symbol]

    if enrichment is not None:
        return Element(symbol=symbol, Z=Z, isotopes=enrichment)

    # Natural abundances from database
    abundances = db.get_natural_abundances(Z)
    # abundances: dict[int, tuple[float, float]] -> A -> (abundance, atomic_mass)
    isotopes = {A: ab for A, (ab, _mass) in abundances.items()}
    return Element(symbol=symbol, Z=Z, isotopes=isotopes)


def resolve_isotopics(
    db: DatabaseProtocol,
    composition: dict[str, float],  # symbol -> mass_fraction OR atom_fraction
    *,
    is_atom_fraction: bool = False,
    overrides: dict[str, dict[int, float]] | None = None,
) -> list[tuple[Element, float]]:
    """Resolve a material composition into (Element, atom_fraction) pairs.

    Args:
        db: Database for natural abundance lookups
        composition: Element symbol -> fraction mapping
        is_atom_fraction: If True, composition values are atom fractions.
            If False, they are mass fractions (will be converted).
        overrides: Optional enrichment overrides per element.
            Keys are element symbols, values are A -> fractional abundance dicts.

    Returns:
        List of (Element, atom_fraction) pairs
    """
    if not is_atom_fraction:
        atom_fracs = mass_to_atom_fractions(composition)
    else:
        atom_fracs = composition

    overrides = overrides or {}
    result: list[tuple[Element, float]] = []

    for symbol, frac in atom_fracs.items():
        enrichment = overrides.get(symbol)
        element = resolve_element(db, symbol, enrichment)
        result.append((element, frac))

    return result


def resolve_formula(
    db: DatabaseProtocol,
    formula: str,
    overrides: dict[str, dict[int, float]] | None = None,
) -> tuple[list[tuple[Element, float]], float]:
    """Resolve a chemical formula into isotopics and compute molecular weight.

    Convenience function combining parse_formula + resolve_isotopics.

    Args:
        db: Database for natural abundance lookups
        formula: Chemical formula (e.g., "MoO3")
        overrides: Optional enrichment overrides per element

    Returns:
        (elements_with_fractions, molecular_weight_u)
    """
    mass_fracs = formula_to_mass_fractions(formula)
    elements = resolve_isotopics(
        db, mass_fracs, is_atom_fraction=False, overrides=overrides,
    )

    # Molecular weight
    elem_counts = parse_formula(formula)
    mol_weight = sum(
        count * STANDARD_ATOMIC_WEIGHT[sym] for sym, count in elem_counts.items()
    )

    return elements, mol_weight

File: frontend/src/lib/compute/materials.ts
/**
 * Material resolution — bridge between formula/composition and HYRR isotopics.
 *
 * Reuses existing utils/formula.ts for parsing, adds isotopic resolution
 * via DatabaseProtocol.
 */

import {
  parseFormula,
  formulaToMassFractions,
  SYMBOL_TO_Z,
  STANDARD_ATOMIC_WEIGHT,
} from "../utils/formula";
import type { DatabaseProtocol, Element } from "./types";

export { parseFormula, formulaToMassFractions, SYMBOL_TO_Z, STANDARD_ATOMIC_WEIGHT };

/** Catalog-entry role — drives category filtering + search ranking. */
export type CatalogRole =
  | "monitor"
  | "target"
  | "window"
  | "structural"
  | "compound"
  | "gas";

/**
 * Known material definitions.
 *
 * `defaultEnrichment` is applied by `resolveMaterial` when no caller-provided
 * `overrides` are given. Explicit overrides always take precedence, per-element.
 */
export interface CatalogEntry {
  density: number;
  massFractions: Record<string, number>;
  defaultEnrichment?: Record<string, Record<number, number>>;
  role?: CatalogRole;
  notes?: string;
}

export const MATERIAL_CATALOG: Record<string, CatalogEntry> = {
  havar: {
    density: 8.3,
    massFractions: {
      Co: 0.42, Cr: 0.20, Ni: 0.13, Fe: 0.184,
      W: 0.028, Mo: 0.02, Mn: 0.016, C: 0.002,
    },
    role: "structural",
    notes: "Cobalt-based alloy, common beam-window material",
  },

  // ─── Monitor foils (natural composition) ───────────────────────────
  natcu: {
    density: 8.96,
    massFractions: { Cu: 1.0 },
    role: "monitor",
    notes: "Natural Cu monitor foil; IAEA-recommended (p,X) monitor reactions",
  },
  natti: {
    density: 4.51,
    massFractions: { Ti: 1.0 },
    role: "monitor",
    notes: "Natural Ti monitor foil; natTi(p,X)48V common monitor",
  },
  natni: {
    density: 8.91,
    massFractions: { Ni: 1.0 },
    role: "monitor",
    notes: "Natural Ni monitor foil",
  },

  // ─── Monoisotopic / simple natural targets ─────────────────────────
  y89: {
    density: 4.47,
    massFractions: { Y: 1.0 },
    role: "target",
    notes: "89Y (monoisotopic natural); 89Y(p,n)89Zr for immunoPET",
  },
  graphite: {
    density: 1.80,
    massFractions: { C: 1.0 },
    role: "window",
    notes: "Reactor-grade graphite (degrader stock). Density varies 1.6–2.26",
  },
  "be-window": {
    density: 1.85,
    massFractions: { Be: 1.0 },
    role: "window",
    notes: "Beryllium cyclotron window. Be metal is toxic — handle per local SOP",
  },

  // ─── Enriched-isotope targets ──────────────────────────────────────
  // Naming convention: `{A}{Symbol}-{role}` avoids collision with the
  // `Element-Mass` free-text form (e.g. "Zn-68") that resolveMaterial
  // strips at line ~188.
  "zn68-electrodeposit": {
    density: 7.13,
    massFractions: { Zn: 1.0 },
    defaultEnrichment: { Zn: { 68: 0.98, 66: 0.015, 67: 0.003, 64: 0.002 } },
    role: "target",
    notes: "68Zn electrodeposit at 98% enrichment; 68Zn(p,n)68Ga",
  },
  "ni64-electrodeposit": {
    density: 8.91,
    massFractions: { Ni: 1.0 },
    defaultEnrichment: { Ni: { 64: 0.95, 62: 0.03, 60: 0.015, 58: 0.005 } },
    role: "target",
    notes: "64Ni electrodeposit at 95% enrichment; 64Ni(p,n)64Cu",
  },
  "mo100-pellet": {
    density: 10.28,
    massFractions: { Mo: 1.0 },
    defaultEnrichment: { Mo: { 100: 0.96, 98: 0.02, 97: 0.01, 96: 0.01 } },
    role: "target",
    notes: "100Mo pressed pellet at 96% enrichment; 100Mo(p,2n)99mTc",
  },
  "ca44-target": {
    density: 1.55,
    massFractions: { Ca: 1.0 },
    defaultEnrichment: { Ca: { 44: 0.97, 40: 0.025, 42: 0.003, 48: 0.002 } },
    role: "target",
    notes: "44Ca target at 97% enrichment; 44Ca(p,n)44Sc for PET",
  },
  "ra226-target": {
    density: 5.50,
    massFractions: { Ra: 1.0 },
    defaultEnrichment: { Ra: { 226: 1.0 } },
    role: "target",
    notes:
      "226Ra target (monoisotopic); α-emitter precursor — licensing + handling" +
      " restrictions apply (226Ra is a high-radiotoxicity source)",
  },
  "th232-target": {
    density: 11.72,
    massFractions: { Th: 1.0 },
    defaultEnrichment: { Th: { 232: 1.0 } },
    role: "target",
    notes:
      "232Th target (monoisotopic natural); route to 225Ac, 99Mo, etc. — export" +
      " controls apply",
  },
  "u238-target": {
    density: 19.05,
    massFractions: { U: 1.0 },
    defaultEnrichment: { U: { 238: 0.9975, 235: 0.0025 } },
    role: "target",
    notes:
      "Depleted U target (~0.25% 235U); route to 99Mo via fission. Export +" +
      " licensing controls apply",
  },
  natu: {
    density: 19.05,
    massFractions: { U: 1.0 },
    defaultEnrichment: { U: { 238: 0.992742, 235: 0.007204, 234: 0.000054 } },
    role: "target",
    notes: "Natural U (0.72% 235U). Licensing + export controls apply",
  },

  // ─── Compounds ─────────────────────────────────────────────────────
  // Natural abundance; mass fractions derived from formula weight.
  baco3: {
    density: 4.29,
    // M = 137.33 + 12.01 + 3*16.00 = 197.34
    massFractions: { Ba: 0.69591, C: 0.06086, O: 0.24323 },
    role: "compound",
    notes: "Barium carbonate; target matrix for enriched-Ba production routes",
  },
  zno: {
    density: 5.61,
    // M = 65.38 + 16.00 = 81.38
    massFractions: { Zn: 0.80340, O: 0.19660 },
    role: "compound",
    notes: "Zinc oxide; compact target form",
  },
  moo3: {
    density: 4.69,
    // M = 95.96 + 3*16.00 = 143.96
    massFractions: { Mo: 0.66657, O: 0.33343 },
    role: "compound",
    notes: "Molybdenum trioxide; pressed-pellet matrix for 100Mo production",
  },
  "h2o-18-enriched": {
    density: 1.11,
    // M = 2*1.008 + 17.999 = 20.015 (enriched water)
    massFractions: { H: 0.10072, O: 0.89928 },
    defaultEnrichment: { O: { 18: 0.97, 16: 0.025, 17: 0.005 } },
    role: "compound",
    notes:
      "Enriched H2[18O] water at 97% 18O; 18O(p,n)18F. Separate key from the" +
      " legacy `H2O-18` compound entry, which stays unenriched",
  },

  // ─── Structural / windows ──────────────────────────────────────────
  "nb-1zr": {
    density: 8.57,
    massFractions: { Nb: 0.99, Zr: 0.01 },
    role: "structural",
    notes: "Niobium–1% zirconium; high-temperature target backing",
  },
  ss316l: {
    density: 7.99,
    massFractions: {
      Fe: 0.654, Cr: 0.17, Ni: 0.12, Mo: 0.025, Mn: 0.02,
      Si: 0.01, C: 0.001,
    },
    role: "structural",
    notes:
      "Austenitic stainless steel, low-carbon grade (midpoint-of-spec" +
      " composition)",
  },
  "inconel-625": {
    density: 8.44,
    // Midpoint-of-range per ASM (pitfall 2); Ni adjusted so sum == 1.000.
    massFractions: {
      Ni: 0.585, Cr: 0.22, Mo: 0.09, Fe: 0.05, Nb: 0.035,
      Al: 0.004, Ti: 0.004, Mn: 0.005, Si: 0.005, C: 0.002,
    },
    role: "structural",
    notes: "Ni-based high-temperature superalloy (midpoint-of-spec composition)",
  },
  "al-6061": {
    density: 2.70,
    massFractions: {
      Al: 0.979, Mg: 0.010, Si: 0.006, Cu: 0.003, Cr: 0.002,
    },
    role: "structural",
    notes: "Aluminium wrought alloy; common target-holder material",
  },
  kapton: {
    density: 1.42,
    // C22H10N2O5 polyimide: M = 22*12.011 + 10*1.008 + 2*14.007 + 5*15.999
    //                        = 264.242 + 10.080 + 28.014 + 79.995 = 382.331
    massFractions: {
      C: 0.69113, H: 0.02637, N: 0.07328, O: 0.20922,
    },
    role: "window",
    notes: "Polyimide film (DuPont Kapton); common thin-window material",
  },
};

/**
 * Merge caller-provided overrides with a catalog entry's `defaultEnrichment`.
 * Per-element: caller overrides win; otherwise fall back to the catalog default.
 */
function mergeEnrichmentWithDefaults(
  defaults: Record<string, Record<number, number>> | undefined,
  overrides: Record<string, Map<number, number>> | undefined,
): Record<string, Map<number, number>> | undefined {
  if (!defaults) return overrides;
  const merged: Record<string, Map<number, number>> = {};
  for (const [sym, abundances] of Object.entries(defaults)) {
    const m = new Map<number, number>();
    for (const [a, f] of Object.entries(abundances)) m.set(Number(a), f);
    merged[sym] = m;
  }
  if (overrides) {
    for (const [sym, m] of Object.entries(overrides)) merged[sym] = m;
  }
  return merged;
}

/** Convert mass fractions to atom fractions. */
export function massToAtomFractions(
  massFractions: Record<string, number>,
): Record<string, number> {
  const moles: Record<string, number> = {};
  let total = 0;
  for (const [symbol, w] of Object.entries(massFractions)) {
    const atomicWeight = STANDARD_ATOMIC_WEIGHT[symbol];
    if (!atomicWeight) continue;
    const m = w / atomicWeight;
    moles[symbol] = m;
    total += m;
  }
  const result: Record<string, number> = {};
  for (const [sym, m] of Object.entries(moles)) {
    result[sym] = m / total;
  }
  return result;
}

/** Resolve an element with natural or enriched isotopic composition. */
export function resolveElement(
  db: DatabaseProtocol,
  symbol: string,
  enrichment?: Map<number, number>,
): Element {
  const Z = SYMBOL_TO_Z[symbol];
  if (Z === undefined) throw new Error(`Unknown element symbol: ${symbol}`);

  if (enrichment) {
    return { symbol, Z, isotopes: enrichment };
  }

  const abundances = db.getNaturalAbundances(Z);
  const isotopes = new Map<number, number>();
  for (const [A, { abundance }] of abundances) {
    isotopes.set(A, abundance);
  }
  return { symbol, Z, isotopes };
}

/** Resolve a material composition into (Element, atom_fraction) pairs. */
export function resolveIsotopics(
  db: DatabaseProtocol,
  composition: Record<string, number>,
  isAtomFraction: boolean = false,
  overrides?: Record<string, Map<number, number>>,
): Array<[Element, number]> {
  const atomFracs = isAtomFraction
    ? composition
    : massToAtomFractions(composition);

  const result: Array<[Element, number]> = [];
  for (const [symbol, frac] of Object.entries(atomFracs)) {
    const enrichment = overrides?.[symbol];
    const element = resolveElement(db, symbol, enrichment);
    result.push([element, frac]);
  }
  return result;
}

/**
 * Resolve a chemical formula into isotopics and molecular weight.
 */
export function resolveFormula(
  db: DatabaseProtocol,
  formula: string,
  overrides?: Record<string, Map<number, number>>,
): { elements: Array<[Element, number]>; molecularWeight: number } {
  const massFracs = formulaToMassFractions(formula);
  const elements = resolveIsotopics(db, massFracs, false, overrides);

  const elemCounts = parseFormula(formula);
  let molWeight = 0;
  for (const [sym, count] of Object.entries(elemCounts)) {
    const w = STANDARD_ATOMIC_WEIGHT[sym];
    if (w) molWeight += count * w;
  }

  return { elements, molecularWeight: molWeight };
}

/** Density estimates for single-element targets (g/cm³). */
export const ELEMENT_DENSITIES: Record<string, number> = {
  H: 0.0899e-3, He: 0.164e-3, Li: 0.534, Be: 1.85, B: 2.34,
  C: 2.26, N: 1.17e-3, O: 1.33e-3, F: 1.58e-3, Ne: 0.900e-3, Na: 0.97,
  Mg: 1.74, Al: 2.70, Si: 2.33, P: 1.82, S: 2.07, Cl: 2.95e-3,
  Ar: 1.78e-3, K: 0.86, Ca: 1.55, Sc: 2.99, Ti: 4.51, V: 6.11, Cr: 7.19,
  Mn: 7.47, Fe: 7.87, Co: 8.90, Ni: 8.91, Cu: 8.96, Zn: 7.13,
  Ga: 5.91, Ge: 5.32, As: 5.73, Se: 4.81, Br: 3.12, Kr: 3.75e-3, Rb: 1.53,
  Sr: 2.63, Y: 4.47, Zr: 6.51, Nb: 8.57, Mo: 10.28, Ru: 12.37,
  Rh: 12.41, Pd: 12.02, Ag: 10.49, Cd: 8.65, In: 7.31, Sn: 7.31,
  Sb: 6.68, Te: 6.24, I: 4.93, Xe: 5.89e-3, Cs: 1.87, Ba: 3.51, La: 6.16,
  Ce: 6.77, Pr: 6.77, Nd: 7.01, Sm: 7.52, Eu: 5.24, Gd: 7.90,
  Tb: 8.23, Dy: 8.54, Ho: 8.80, Er: 9.07, Tm: 9.32, Yb: 6.57,
  Lu: 9.84, Hf: 13.31, Ta: 16.65, W: 19.25, Re: 21.02, Os: 22.59,
  Ir: 22.56, Pt: 21.45, Au: 19.30, Hg: 13.55, Tl: 11.85,
  Pb: 11.34, Bi: 9.78, Ra: 5.50, Th: 11.72, U: 19.05,
};

/** Compound density estimates (g/cm³). */
export const COMPOUND_DENSITIES: Record<string, number> = {
  H2O: 1.0, "H2O-18": 1.11, MoO3: 4.69, Al2O3: 3.95,
};

/** Optional custom density override lookup — set by the UI layer. */
let customDensityLookup: ((formula: string) => number | null) | null = null;

/** Register a function that returns custom material densities. */
export function setCustomDensityLookup(fn: (formula: string) => number | null): void {
  customDensityLookup = fn;
}

/** Custom material composition lookup — returns mass fractions if available. */
let customCompositionLookup: ((identifier: string) => Record<string, number> | null) | null = null;

/** Register a function that returns custom material mass-fraction compositions. */
export function setCustomCompositionLookup(fn: (identifier: string) => Record<string, number> | null): void {
  customCompositionLookup = fn;
}

/**
 * Resolve a material identifier (name, formula, or element symbol with mass number)
 * into elements, density, and molecular weight.
 */
export function resolveMaterial(
  db: DatabaseProtocol,
  identifier: string,
  overrides?: Record<string, Map<number, number>>,
): { elements: Array<[Element, number]>; density: number; molecularWeight: number } {
  // Check material catalog first (named alloys)
  const lowerIdent = identifier.toLowerCase();
  const catalogEntry = MATERIAL_CATALOG[lowerIdent];
  if (catalogEntry) {
    const effective = mergeEnrichmentWithDefaults(catalogEntry.defaultEnrichment, overrides);
    const elements = resolveIsotopics(db, catalogEntry.massFractions, false, effective);
    return { elements, density: catalogEntry.density, molecularWeight: 0 };
  }

  // Check for custom material with stored mass fractions (wt% materials)
  if (customCompositionLookup) {
    const massFracs = customCompositionLookup(identifier);
    if (massFracs) {
      const elements = resolveIsotopics(db, massFracs, false, overrides);
      let density: number | undefined;
      if (customDensityLookup) {
        const d = customDensityLookup(identifier);
        if (d !== null) density = d;
      }
      if (density === undefined) density = 5.0;
      return { elements, density, molecularWeight: 0 };
    }
  }

  // Strip mass numbers from element identifiers: "Mo-100" → "Mo", "Ra-226" → "Ra"
  const formulaClean = identifier.replace(/-\d+/g, "");

  // Parse as chemical formula
  const { elements, molecularWeight } = resolveFormula(db, formulaClean, overrides);

  // Determine density
  let density: number | undefined;

  // Check custom materials first
  if (customDensityLookup) {
    const custom = customDensityLookup(identifier) ?? customDensityLookup(formulaClean);
    if (custom !== null) density = custom;
  }

  if (density === undefined) {
    if (COMPOUND_DENSITIES[identifier]) {
      density = COMPOUND_DENSITIES[identifier];
    } else if (COMPOUND_DENSITIES[formulaClean]) {
      density = COMPOUND_DENSITIES[formulaClean];
    } else {
      // Single-element target: use element density
      const parsed = parseFormula(formulaClean);
      const symbols = Object.keys(parsed);
      if (symbols.length === 1 && ELEMENT_DENSITIES[symbols[0]]) {
        density = ELEMENT_DENSITIES[symbols[0]];
      } else {
        // Fallback: rough estimate from mass fractions
        density = 5.0;
        console.warn(`[materials] No density for "${identifier}", using default ${density} g/cm³`);
      }
    }
  }

  return { elements, density, molecularWeight };
}
